### PR TITLE
Add converter for downstream response

### PR DESCRIPTION
**Pull request by jaygehlot**
Add converter for downstream response

### CODE REVIEW

File: orion/src/main/java/com/marksandspencer/downstream/Converter.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package com.marksandspencer.downstream;
+
+public interface Converter<F, T> {
```
Comment (clarkster) [inline orion/src/main/java/com/marksandspencer/downstream/Converter.java]: Any reason not to use the built in `java.util.function.Function` for this?

File: orion/src/main/java/com/marksandspencer/downstream/DownstreamHandler.java [hunk 2]
```diff
@@ -90,8 +104,8 @@ private ResponseEntity handleClientError(HttpClientErrorException exception) {
         } catch (Exception e) {
             // ignore
         }
-        ErrorMessage msg = parsedMessage.orElseGet(
-                        () -> new ErrorMessage(exception.getStatusCode().getReasonPhrase(), exception.getMessage()));
+        ErrorMessage msg = parsedMessage.orElse(
+                       new ErrorMessage(exception.getStatusCode().getReasonPhrase(), exception.getMessage()));
```
Comment (clarkster) [inline orion/src/main/java/com/marksandspencer/downstream/DownstreamHandler.java]: I guess it doesn't matter a lot but the use of `orElseGet` and a Supplier was deliberate to save the cost of constructing the `ErrorMessage` when it isn't needed.

https://blog.jetbrains.com/idea/2016/07/java-8-top-tips

File: orion/src/main/java/com/marksandspencer/downstream/Converter.java (hunk 1 above)
Comment (yasinefe) [inline orion/src/main/java/com/marksandspencer/downstream/Converter.java]: You are right, I have changed to use Function

File: orion/src/main/java/com/marksandspencer/downstream/DownstreamHandler.java (hunk 2 above)
Comment (yasinefe) [inline orion/src/main/java/com/marksandspencer/downstream/DownstreamHandler.java]: Wooow, it is a really tricky one. I did not know and could not think. By the way, in the other case JVM will create Supplier object to pass it to orElseGet. But, It must be performant than creation error message.

Anyway, it is better to keep in mind this and apply always orElseGet. I have reverted back.

Comment (clarkster) [inline orion/src/main/java/com/marksandspencer/downstream/DownstreamHandler.java]: Although the disadvantage of the `orElseGet` version is that its not as simple or readable :-) Don't know which is best really, I suppose in general readability beats micro-optimising

**Comment by clarkster** [comment]
looks good to me - just needs a conflict resolving

**Pull request merged**